iOS Support

dreautall opened the pull request:
iOS Support.

I currently have no plans to get an Apple Developer account and officially publish to the store due to the associated cost, but I nonetheless want to potentially support iOS now that I have the capability to debug for it (thanks to a Macbook I use as my primary device now).

This is a "side-task", which doesn't have the highest priority for now. However, any help is appreciated, especially in finding or fixing UI issues!

Tasks:
- [x] Basic iOS support to allow the app to run
- [x] Adaptive Widgets where provided by Flutter
- [x] Remove Notification Listener mentions
- [x] Check if additional adaptive widgets, for example with https://pub.dev/packages/flutter_platform_widgets, are worth the effort
- [x] Check for UI errors:
  - [x] AlertDialog buttons look completely off

dreautall committed:
[ios] Initial Waterfly III iOS implementation

dreautall committed:
[ios] Initial Waterfly III iOS implementation

dreautall committed:
[ios] general fixes to allow app launch

dreautall committed:
[ios] general fixes to allow app launch

dreautall committed:
[ios] fix image_picker

dreautall committed:
[ios] fix image_picker

dreautall committed:
Merge branch 'develop' into feature/ios

dreautall committed:
Merge remote-tracking branch 'origin/feature/ios' into feature/ios

dreautall committed:
Merge branch 'develop' into feature/ios

dreautall committed:
update xcode

dreautall committed:
make all potential adaptive widgets adaptive

bigger change for ListBuilder which uses a hardcoded CircularProgessIndicator.

dreautall committed:
[settings] hide notification listener on iOS

dreautall committed:
Add "(Android only)" disclaimer for notification listener

dreautall committed:
Merge branch 'develop' into feature/ios

# Conflicts:
#	lib/notificationlistener.dart
#	lib/pages/home/transactions.dart
#	lib/pages/home/transactions/filter.dart
#	lib/pages/settings.dart

dreautall committed:
upgrade to ios files from latest flutter version

dreautall committed:
run on latest ios 26.1

dreautall committed:
only load notifications for Android

mskarbek commented:
After the first quick test, the most obvious problem is the lack of a filters widget on the transactions page:
<img width="1206" height="2622" alt="Simulator Screenshot - iPhone 17 - 2025-11-14 at 06 55 55" src="https://github.com/user-attachments/assets/0ad2262c-dd59-4fc1-8ebd-72eca7d3e221" />
Please let me know what kind of help (if any) you would like with iOS support work. I'm not a Flutter developer, but I can test, and I could learn a thing or two about Dart and Flutter with this project.

dreautall commented:
Hey, thank you for wanting to help!

I think the main issue isn't really coding (well, not yet). It's actually how I want to approach it, and as I don't use iOS daily I'm a little bit lost. Basically:

1. Theme the app for iOS (so that it looks like a native app)
2. Keep the current app theme (Material Design, the "Android look").

Googling around, both methods seem to have supporters. Apparently, Google for the most part also keeps Material Design in their apps and doesn't re-theme the app to loo more "native". As a quick example (all from the iOS simulator, please excuse some stuff not looking properly):

1. Fully Material Design (looks like the app on Android): Easy to do.
<img width="347" height="447" alt="Screenshot 2025-11-14 at 08 48 29" src="https://github.com/user-attachments/assets/28a7bbb9-10ef-4aef-9ca6-10ec76a3d53a" />

2. Themed Design (looks like iOS): Not really easy to accomplish, as it duplicates a lot of code.
<img width="287" height="366" alt="Screenshot 2025-11-14 at 08 56 25" src="https://github.com/user-attachments/assets/34ae626b-8d96-44a5-bad6-ca739a1c998f" />

3. Do just the minimum adaptions as outlined in https://m2.material.io/design/platform-guidance/cross-platform-adaptation.html#cross-platform-guidelines (for Material Design 2, not 3 - there is no equivalent for Material Design 3). This basically shifts the most common elements to where they are expected on iOS (centered headings, different back buttons etc.)


If you are regularly using iOS (or even both Android & iOS), any insight would be appreciated!

(What throws the errors in your screenshot is actually me trying to do a mix: Have the dialog look like iOS, but all buttons/Text fields like Material Design. That obviously technically doesn't work properly).

mskarbek commented:
Soon I will switch to iOS as a primary phone; that is why I took interest in this PR. I was already using iOS some time ago, and I didn't mind the Material themeing of applications. There is no need to force native design; for me, it's much more important to keep functionality (sadly, notification listening won't be possible).
IMO, making the app fully usable on iOS would be the primary goal. Adaptations and theme adjustments should be the next stage if there is real interest in native feel among other users. 
Edit: One way of making an app more "native" instead of theming would be using `cupertino_http` on `Platform.isIOS` instead of switching to a basic http client (https://github.com/dreautall/waterfly-iii/pull/595/files#diff-079afea380619a7377576293dfe986abcb99f78636d247b16ae7b14457baef4eR100).

dreautall committed:
always show Android style AlertDialogs

prevents issues with buttons etc. not loading due to missing `Material` ancestor

dreautall committed:
fix format errors

dreautall committed:
use `cupertino_http` as HTTPClient

dreautall commented:
Thanks for the hint with `cupertino_http`, implemented that!

Otherwise all errors should now be fixed, with the app (mostly) looking like the Android one, except for a couple of "easy" stuff like toggle switches etc.

dreautall committed:
[ios] splash screen

dreautall committed:
[ios] fix add shortcut action icon

mskarbek commented:
<img width="1206" height="343" alt="over" src="https://github.com/user-attachments/assets/b83fa3f6-5b50-49ca-803a-f8b56eaae7d7" />

```
======== Exception caught by rendering library =====================================================
The following assertion was thrown during layout:
A RenderFlex overflowed by 25 pixels on the right.

The relevant error-causing widget was: 
  Row Row:file:///Users/marcin/Projects/waterfly-iii/lib/pages/home/main.dart:1500:32
The overflowing RenderFlex has an orientation of Axis.horizontal.
The edge of the RenderFlex that is overflowing has been marked in the rendering with a yellow and black striped pattern. This is usually caused by the contents being too big for the RenderFlex.

Consider applying a flex factor (e.g. using an Expanded widget) to force the children of the RenderFlex to fit within the available space instead of being sized to their natural size.
This is considered an error condition because it indicates that there is content that cannot be seen. If the content is legitimately bigger than the available space, consider clipping it with a ClipRect widget before putting it in the flex, or using a scrollable container rather than a Flex, like a ListView.

The specific RenderFlex in question is: RenderFlex#847b6 relayoutBoundary=up20 OVERFLOWING
...  parentData: offset=Offset(12.0, 12.0) (can use size)
...  constraints: BoxConstraints(0.0<=w<=354.0, 0.0<=h<=Infinity)
...  size: Size(354.0, 24.0)
...  direction: horizontal
...  mainAxisAlignment: spaceBetween
...  mainAxisSize: max
...  crossAxisAlignment: center
...  textDirection: ltr
...  verticalDirection: down
...  spacing: 0.0
◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤◢◤
====================================================================================================
```
Behavior observed when switching to the `pl` language in the `Category Summary for current month` widget.

dreautall commented:
Would there be a shorter translation for that string?

mskarbek commented:
I have submitted a shorter translation in Crowdin.

dreautall committed:
Merge branch 'develop' into feature/ios

dreautall committed:
Merge branch 'develop' into feature/ios

dreautall committed:
Merge branch 'develop' into feature/ios

dreautall committed:
fix sharing to app with latest `flutter_sharing_intent` version

dreautall commented:
Bookmark (for myself) for auto-building the app with a github action: https://medium.com/team-rockstars-it/the-easiest-way-to-build-a-flutter-ios-app-using-github-actions-plus-a-key-takeaway-for-developers-48cf2ad7c72a

Still not sure if I want to "invest" the recurring payment for an Apple Dev profile though :(
